src/OrbitGl/TimerTrack.cpp

```diff
@@ -103,34 +128,100 @@ void TimerTrack::UpdatePrimitives(uint64_t min_tick, uint64_t max_tick,
+              color[2] = static_cast<unsigned char>(static_cast<float>(color[2]) * 0.8f);
+            }
+
+            double intersection_start_us = start_us;
+            start_us = time_graph_->GetUsFromTick(prev_timer_info.end());
+
+            WorldXInfo world_x_info = ToWorldX(intersection_start_us, start_us, inv_time_window,
+                                               world_start_x, world_width, canvas->GetWidth());
+
+            if (world_x_info.is_visible_width && !is_collapsed) {
+              Vec3 top_left(world_x_info.world_x_start, world_timer_y + box_height, z);
+              float world_intersection_end_x =
+                  world_x_info.world_x_start + world_x_info.world_x_width;
+              Vec3 bottom_right(world_intersection_end_x, world_timer_y, z);
+              Vec3 top_right(world_intersection_end_x, world_timer_y + box_height, z);
+              Triangle triangle(top_left, bottom_right, top_right);
+              auto user_data = std::make_unique<PickingUserData>(
+                  &text_box, [&](PickingId id) { return this->GetBoxTooltip(id); });
+              batcher->AddTriangle(triangle, color, std::move(user_data));
+            }
+          }
+        }
+
+        // Check if the next timer overlaps with the current one, and if so draw the overlap
+        // as triangles rather than as overlapping rectangles.
+        if (k + 1 < block.size()) {
+          const TimerInfo& next_timer_info = block[k + 1].GetTimerInfo();
+          // We also compare the type, as for the Gpu timers, timers of different type but same
+          // depth are drawn below each other (and thus do not overlap).
+          if (timer_info.end() > next_timer_info.start() &&
+              next_timer_info.type() == timer_info.type()) {
+            double intersection_end_us = end_us;
+            end_us = time_graph_->GetUsFromTick(next_timer_info.start());
+
+            WorldXInfo world_x_info = ToWorldX(end_us, intersection_end_us, inv_time_window,
+                                               world_start_x, world_width, canvas->GetWidth());
+            if (world_x_info.is_visible_width && !is_collapsed) {
+              Vec3 bottom_left(world_x_info.world_x_start, world_timer_y, z);
+              Vec3 bottom_right(world_x_info.world_x_start + world_x_info.world_x_width,
+                                world_timer_y, z);
+              Vec3 top_left(world_x_info.world_x_start, world_timer_y + box_height, z);
+              Triangle triangle(bottom_left, bottom_right, top_left);
+
+              auto user_data = std::make_unique<PickingUserData>(
+                  &text_box, [&](PickingId id) { return this->GetBoxTooltip(id); });
+              batcher->AddTriangle(triangle, color, std::move(user_data));
+            }
+          }
+        }
+
+        // If the timer is overlapping completely with the previous and next timer, we neither draw
+        // text nor a box.
+        if (start_us > end_us) continue;
```

Comment:
How many of these are we having? Are them important? Are you planning to draw them in the future?

I would prefer to draw them without text and then having a pop up when you hover the mouse there, because the user could be interested.

Reply:
We should discuss about this. I have a sense there is a way that the overall algorithm could be unified and lead to the result that @karupayun is proposing. 

Every box has a start and end time on the top of its depth (start_top, end_top) and a start and end time on the bottom of its depth (start_bottom, end_bottom). In the non-overlapping case start_top == start_bottom == start and end_top == end_bottom == end. 

In the general case, start_top == start and end_bottom == end always is true. And end_top is always the minimum of end and start(next), and start_bottom is always the maximum of start and end(previous). 

You always need to draw the two triangles (start_top, end_bottom, end_top) and (start_top, start_bottom, end_bottom) if traversing counter-clockwise. 

This would lead to timers that are "squashed" between two other timers to be drawn as a slanted quadrangle that fits in between the others. 

The assumption I'm making is that one command buffer / debug label execution cannot "overtake" another one. That is, if start < start(next), then also end < end(next). 

(The text label is a problem though. I assume that's why you need to draw a quad in the "middle" and the other two triangles?)

Reply:
We discussed this yesterday a bit offline, and this is exactly what I prototype right now. Though, I am not sure, if we want to keep the rectangle as a base case for non-overlaps, as we then need to draw less elements (might be better in perfomance). I can do some experiments.

Reply:
As discussed offline yesterday, I agree with Ronald. I should have written something here, sorry for that.

The 2 triangles proposals works for every case, as Ronald pointed out. But we should avoid it in cases where it's clear that there isn't any intersection for performance.

As Ronald also pointed out, we spoke that we should not print at all timers who are inside of others. (t1_start < t2_start and t1_end > t2_end). With this assumption, we can easily draw the parallelograms.

But there is still a little issue with coloring. We used the shadowBoxes in Timers to differentiate between consecutive timers, and this is what it is causing problems here. So, how to draw the triangles to:
1- Make the feeling that they below to the same parallelogram and
2- Differentiate with the previous/next.?

We have this n%2 issue for that but we are assuming that events come in order. I think that the proposal solution is inserting them with this n%2 algorithm but repainted them after, so when finalizing the capture, we have the invariant: "for every consecutive pair of timers, one is darker and one is lighter". Am I right @florian-kuebler or we still have to discuss that?

Reply:
Done. We are now drawing always two triangles. I'll do the performance experiments tomorrow.